### ISSUE TITLE
Need help with match_template!

### ISSUE THREAD

**Issue by khgkim** [twistedfall/opencv-rust#168]
Hi,

I've been struggling with how **match_template** works for the past couple days and would really appreciate some help here...

This is the code I have so far:

    // Read images
    let template = imgcodecs::imread(SAMPLE_TEMPLATE, imgcodecs::IMREAD_COLOR).unwrap();
    let image = imgcodecs::imread(SAMPLE_SOURCE, imgcodecs::IMREAD_COLOR).unwrap();

    // Resize template
    let mut reduced = Mat::default().unwrap();
    imgproc::resize(
        &template, 
        &mut reduced, 
        core::Size {
            width:0, 
            height:0
        }, 
        0.5, 
        0.5, 
        imgproc::INTER_AREA 
    ).unwrap();

    // Optional
    // Only two matching methods currently accept a mask: TM_SQDIFF and TM_CCORR_NORMED
    let mask = Mat::default().unwrap();

    // Construct the result matrix, a single-channel 32-bit floating-point.
    // If image is W x H and template is w x h, then result is (W - w + 1) x (H - h + 1)
    let zero = core::Mat::zeros(
        image.rows() - reduced.rows() + 1,
        image.cols() - reduced.cols() + 1,
        core::CV_32FC1,
    )
    .unwrap();

    let mut result = zero.to_mat().unwrap();

    // Match template
    let temp = imgproc::match_template(
        &image, // input image to be searched, 8U or 32F, size W x H
        &reduced, // template to use, same type as 'image', size w x h
        &mut result, // result image, type 32F, size (W - w + 1) x (H - h + 1)
        imgproc::TM_CCOEFF_NORMED, 
        &mask // optional mask
    );

    // Find the location of the best match
    let mut min_val: f64 = 0.0;
    let mut max_val: f64 = 0.0;
    let mut min_loc: core::Point = core::Point::new(0,0);
    let mut max_loc: core::Point = core::Point::new(0,0);
    let mask = Mat::default().unwrap();

    core::min_max_loc(&result, &mut min_val, &mut max_val, &mut min_loc, &mut max_loc, &mask); 

    // imgproc::rectangle(img: &core::Mat, rec: core::Rect, color: core::Scalar, thickness: i32, line_type: i32, shift: i32)

    imgcodecs::imwrite("test_img.jpg", &result, &opencv::types::VectorOfi32::new()); 

Basically, this is how I'm approaching match_template:

1. Read in the template and source images
2. Resize the template image because it can't be larger than the source
3. Create an empty mask because it's optional for our comparison method
4. Construct the result matrix for match_template ((W - w + 1) x (H - h + 1))
5. After calling match_template, find the coordinates of the best match using core::min_max_loc.

The goal is to save the detected image, but I'm not sure how to achieve this with the code that I have. Am I missing any step(s) here? 

Thanks for the help!

**Comment by twistedfall** [comment]
Hi, sorry for the delay, I don't appear to have too much free time. From your description I don't seem to understand what exactly is the problem? Only saving the image? The last line should work fine with that. What result are you getting?

**Comment by khgkim** [comment]
No worries.

The problem is that calling match_template() always results in a black image, when the goal is to save an image with the detected template. I'm not sure how to get the coordinates of the detected image.

**Comment by twistedfall** [comment]
The result of `match_template` is a matrix of weights, it's not an image, basically you get a lot of numbers like this:
```
...
        -0.015997173,
        -0.011935301,
        -0.0076606544,
        -0.0035734226,
        0.0001562627,
        0.003847433,
        0.0076474943,
        0.011895044,
...
```
So there is no point in saving it as image. You can dump this `Mat` contents with:
```rust
dbg!(&result.data_typed::<f32>());
```
Generally you should go over the results in that `Mat` and find poinnts with high enough scores for you, those will signify the places where template matched with high enough probability. I would also advise to seek the tutorials on OpenCV template matching, they will most probably be in C++ or Python, but are quite easy to adapt to Rust.
